### DISCUSSION TITLE
Share your use cases for this project

### DISCUSSION THREAD

**Discussion by rednblkx** [rednblkx/HomeKey-ESP32#17] [Show and tell]
I see people with various use cases or ideas of use cases like a PoE powered ESP32, integration in ESPHome, home key actioned relay module and I want to hear more about it which is the reason this discussion was created.
You can share here how you plan on using this Home Key thing and ESP32 or how are you currently using this whole thing.

**Comment by tonysprenk** [comment]
Would be great to have a native HomeKit implementation with a relay. I have a solenoid lock on my garden gate which is controlled by an esp8266 relay board. I would love to swap out this board with an esp32 variant and pair it with an NFC reader. Hopefully this project can accommodate my setup in the future. My coding knowledge is not proficient enough to make this happen myself unfortunately. [HAA](https://github.com/RavenSystem/esp-homekit-devices)  has recently  implemented support for esp32 so perhaps inspiration could be drawn from there. Thanks for the great project!

**Reply by rednblkx** [reply]
I'm still trying to figure out the structure for the whole thing, and GPIO Control is something i want to include in a future revision once i figure out how everything should come together as i'm want to make everything extremely modular. 
Right now you should be able to just replace the `mqtt.publish(MQTT_STATE_TOPIC` lines with your relay implementation.
If you need help, you can email me.

**Reply by 5and0** [reply]
Nice one ...  I am going to use this purely for the reader and manage the lock control from a secure device. I would like to send visual confirmation to an LED and/or buzzer once authorised / unlocked.

**Comment by 011V32** [comment]
My plan is to place the NFC reader on the outside of the door (above the bell and the light switch behind a panel) and use it to open a wireless lock (Nuki). Ideally, this works not only with HomeKey but also with AirTags... (So that the children can also open the door)
Whether I use ESPHome, HomeAssistant or something else is irrelevant to me. (Ideally it is a standalone function and additional systems such as HomeAssistant only access information from it and can control it...)
What is still open for me: how long can the cable between ESP32 and PN532 be? (This depends on where I have to place my ESP32. But in any case, I use an Ethernet (with PoE) variant.
In any case, a big thank you for your efforts! I've been waiting for this for years :)

**Reply by rednblkx** [reply]
I'm intrigued by the AirTags thing, what would be the flow for that?
As for the cable length, not sure but i think it is whatever makes the SPI protocol work, also the base library HomeSpan connects to network via Wifi and i don't think Ethernet would override this, though i don't have an Ethernet module so i might be wrong :man_shrugging: but i guess it would still be useful for PoE.

**Reply by thorsten-gehrig** [reply]
AirTags do have an NFC chip - and therefore it do work as simple access key like any other NFC card!
I use those since a long time :-)
Only downside: you must use them with the right side (the top side) - as their NFC antenna is not as strong to go through the battery and battery case... nevertheless: it´s a great 2-in-1 (as the airtags itself have already a great value - the key is just added).
And also good: no extra implementation is necessary! :-)
Best
Thorsten

**Reply by rednblkx** [reply]
huh, interesting, did not knew that albeit never used them but that is pretty cool, thanks.

**Reply by petkoe** [reply]
Hi Thorsten, in what context do you use the airtags now i.e. what hardware to execute home bridge/knx ?

regards
Peter

**Reply by thorsten-gehrig** [reply]
Hi
Currently I´m having an NFC reader connected to an Arduino which is connected to my KNX bus. I´ll list read the NFC-Tag (either from cards, RFID-keyfobs or AirTags) to identify the tag - and open the door (via the KNX)
This is a 10 years old project which is running since then - also based on an PN532.
Code is listed here: https://knx-user-forum.de/forum/öffentlicher-bereich/knx-eib-forum/diy-do-it-yourself/35573-arduino-rfid-knx-anbindung  (sorry - it´s in German - but the code is super-simple - in the last post from 2021 with a bug-fix :-))
Kind regards
Thorsten

**Reply by petkoe** [reply]
Hi Thorsten, thats were i know your name from. I am also German and in the forum (pekoe). I am no hardware guy and more high level software but will try to start with this project and also the Arduino from your project. Schönen Sonntag Peter

**Comment by 5and0** [comment]
I am looking to replace the following with this implementation. As I mentioned earlier, I would like to add the visual (led) and audible (buzzer) confirmation that I can control once an authorised tag has been read by home assistant. 

https://github.com/adonno/tagreader

Thanks @rednblkx awesome work on this, something I have been waiting for since homekey was announced.

**Comment by 011V32** [comment]
From your perspective, what is the standard procedure for HomeKey functionality?
(I imagine this to be different for locks that can only lock and unlock than for locks that can also unlatch.)
For locks with lock and unlock, it probably makes sense to change the state after each authentication. For locks that can also be unlatched, it probably only makes sense to unlatched the lock after each authentication, since you don't know which of the other 2 states you want to get into. What are your thoughts here?

**Reply by tonysprenk** [reply]
The point of home key is to only get into the residence so it shouldn’t matter what state the door lock is in it should always open the door. The lock has sensors built in to know when to stop turning. So when it’s fully locked it will turn the lock longer to open the door than when it’s only latched.

**Reply by rednblkx** [reply]
It seems i should touch upon more on this on the first page of the wiki as i've seen it pop up a few times.
The project currently fully relies on MQTT to have its state "in sync" which means it falls onto the user to handle the appropriate state updates of the lock itself and the homekit entity therefore you are able to translate actions. 

However, i did encounter a situation in #38 and subsequently implemented a couple changes in relation to this that are available in the dev branch but i haven't yet merged it as i haven't prepared an update for the wiki yet. 

Essentially the changes remove that publication of state on each authentication and have introduced custom mqtt topics with custom states and actions, this way hopefully it covers more use cases and edge cases(like the one in the original issue).

**Comment by aescolastico** [comment]
i have a potential use case for giving my mailbox homekey support with an electric drop bolt lock.

**Comment by x23piracy** [comment]
Simply with a Nuki https://forum.smartapfel.de/forum/thread/9036-apple-home-key/?postID=149903#post149903

**Comment by eigenphase** [comment]
How I plan to use it: Put the reader at my front door. When card scanned successfully, event in Home Assistant unlocks the door. Surprising, right? 😉

But a few things which I'd love to implement and need to see how they pan out:

1. Ideally the reader is behind the stucco wall. I have access to the sheathing but it's still a bit wood and maybe an inch of stucco. Might be too much attenuation. In this case, re-try the same with behind the door (1" of wood)
2. If not, I envision only placing the reader part in a thin, custom box under the doorbell and bringing the SPI behind the wall. Would be a few centimeters. There the ESP, ideally PoE connected
3. I am planning to drive an LED and piezo speaker to signal success/not success: Green when ready and flashing/chime when access OK. Red flashing (and different chime) when did not succeed
4. I don't care about Apple Home so would only use it to set it up. But I also want to share the key with family members. Right now, this needs AppleTV/HomePod as "hub". Would be great if there are any other solutions without needing the Apple stuff
5. The ultimate great thing would be to wire it directly to my door lock. Mine is a Kwikset 914. This requires GPIO pads that operate the lock, possibly due to debug functionality implemented by Kwikset. Unfortunately I have not found anything. This would avoid the delay for Wifi, Zigbee, Z2M and HA.

**Reply by rednblkx** [reply]
It's possible you might not need the hub, issue only lies with the "new homekit architecture" which requires a home hub in the first place to upgrade otherwise you are with the old and at least on older ios versions it allowed adding people to the Home, don't know if they changed that ever since.

If you can invite people to your Apple Home, i think that should be enough.

> The ultimate great thing would be to wire it directly to my door lock. Mine is a Kwikset 914. This requires GPIO pads that operate the lock, possibly due to debug functionality implemented by Kwikset. Unfortunately I have not found anything. This would avoid the delay for Wifi, Zigbee, Z2M and HA.

On a brief search i couldn't find anything about the GPIO pads for this lock, if you can point me to something about this i'll look into it though if it's something very specific probably won't implemented into this project with emphasis on this. In the meantime, there is a rudimentary HIGH-LOW One Pin thing present in the latest release under "GPIO Actions" but i assume that won't be enough.

**Comment by electrontube** [comment]
For my use case, the ESP-Home Reader only needs to take care of the NFC communication, there is no access logic on the ESP itself. For each successfully read device, a unique ID is sent via MQTT to a central unit, which contains the entire access logic (such as time-dependent access profiles). The central instance evaluates the UID, checks the associated profile and sends corresponding triggers for actions to the ESP via MQTT.  It should therefore be possible to configure 2-3 outputs on the ESP and link them to MQTT topics. In addition, 1-2 inputs that publish an MQTT topic would be useful.

There should be TLS and at least some basic password protection for the ESP web interface.

Since the ESP is inevitably located near the door, it makes sense to operate things like door openers, light switches, doorbells, etc. directly with GPIOs on this ESP, which would make additional hardware (another ESP) obsolete.

I am only concerned here with the ESP part, the central components and logic as well as the interfacing to the actuators (e.g. relays for door openers) are not the subject of this project.

**Comment by fuelle** [comment]
I wanna use it to activate the electric door opener for a gate (and house entrance door) of a apartment building.

One reader at the gate, a second one on the house entrance door. Both are always closed, but never locked. I would prefer to display them as locked in HomeKit, temporarily "unlock" them and then go back to locked state.

I got two sets of ESP32, PN532 and a relay. The relay can be be connected to the existing door opener wires. 
No PN532 yet, but I already turn on/off a bulb via the Home app. I often experience a delay of multiple seconds. I guess thats caused by WiFi. 
My assumption is, that it will be much faster if the relay is triggered by the NFC reader. Correct?

Seems like the version from @Maxime0506 would fit my needs. But there is no binary, and I wasn't able to build one.

@rednblkx Any idea when HomeKey could become part of ESPHome? Or is there a chance you might add a delay for the GPIO actions?

Thanks for the project!

**Reply by Maxime0506** [reply]
Hello @fuelle 
I wanted to inform you that I just reprogrammed the new update version that was recently published by @rednblkx so it now also works without MQTT, and if you want, I can help you with your project. Feel free to contact me I will update my GitHub page soon with the updated version.

**Reply by fuelle** [reply]
Thanks, for now a compiled bin file would help me a lot. Couldn't figure out how to do that.

**Reply by rednblkx** [reply]
@fuelle i've actually recently created a HomeKit component to be used on ESPHome and that of course includes HomeKey and i'm now working on documentation, if you are interested, give it a go and let me know what breaks 🤪

Also a "state rebound" type of functionality is in the books to be added in probably the next version which will let you define if you want the state for GPIO Actions to be static or momentary.

**Reply by rednblkx** [reply]
https://github.com/rednblkx/HAP-ESPHome

You can find there example yaml files

No idea if it can ever become part of the official ESPHome

**Comment by thanhtoan009** [comment]
You should add a gpio pin for the proximity sensor or magnetic sensor, it will recognize when the door reaches the closing point, after 0.5s 1s it will close the mechanical lock and lock on the homekit.
